token-tracker/src/truth-sources/token-grant.js

```diff
@@ -0,0 +1,142 @@
+    super(context, finalBlock)
+  }
+
+  async initialize() {
+    const TokenGrant = new Contract(TokenGrantJSON, this.context.web3)
+
+    this.context.addContract("TokenGrant", TokenGrant)
+
+    this.tokenGrant = await this.context.getContract("TokenGrant").deployed()
+  }
+
+  /**
+   * @typedef {Object} TokenGrant
+   * @property {string} id
+   * @property {Address} grantee
+   * @property {BN} balance
+   */
+
+  /**
+   * Gets all token grants based on TokenGrantCreated event of TokenGrant contract.
+   * Returns details of each grant like ID, grantee and balance. The balance is
+   * calculated at the pre defined final block as total grant amount minus withdrawn
+   * and revoked tokens. The balance includes tokens that were staked, it's important
+   * to exclude the stake amount in TokenStaking source of truth to avoid value
+   * duplication.
+   * @return {TokenGrant[]} Token grant with ID, grantee and balance at final block.
+   */
+  async findTokenGrants() {
+    logger.info(
+      `looking for TokenGrantCreated events emitted from ${this.tokenGrant.options.address} ` +
+        `between blocks ${this.context.deploymentBlock} and ${this.finalBlock}`
+    )
+
+    /** @type {Array} */
+    const events = await getPastEvents(
+      this.context.web3,
+      this.tokenGrant,
+      "TokenGrantCreated",
+      this.context.deploymentBlock,
+      this.finalBlock
+    )
+    logger.info(`found ${events.length} token grant created events`)
+
+    /** @type {String[]} */
+    const tokenGrantIDs = events.map((event) => event.returnValues.id)
+
+    /** @type {TokenGrant[]} */
+    const tokenGrants = []
+    for (const id of tokenGrantIDs) {
+      const grant = await callWithRetry(
+        this.tokenGrant.methods.getGrant(id),
+        undefined,
+        undefined,
+        this.finalBlock
+      )
+
+      // Balance is the total grant amount minus revoked and withdrawn tokens. The
+      // value includes staked tokens.
+      const balance = toBN(grant.amount)
+        .sub(toBN(grant.withdrawn))
+        .sub(toBN(grant.revokedAmount))
+
+      tokenGrants.push({ id: id, grantee: grant.grantee, balance: balance })
+    }
+
+    dumpDataToFile(tokenGrants, TOKEN_GRANT_DUMP_PATH)
+
+    return tokenGrants
+  }
+
+  /**
+   * Converts list of grants to a map of owners and their balances. It resolves
+   * owner for each grantee and combines balances for the same owners.
+   * @param {Array<TokenGrant>} tokenGrants Stake owners to filter.
+   * @return {Map<Address,BN>} Map of owners and their balances.
+   */
+  async resolveOwnersBalances(tokenGrants) {
+    /** @type {Map<Address,BN>} */
+    const ownersBalances = new Map()
+
+    for (const tokenGrant of tokenGrants) {
+      const { grantee, balance } = tokenGrant
+
+      const owner = await resolveGrantee(this.context.web3, grantee)
+
+      if (owner !== grantee)
+        logger.debug(`resolved owner of grantee [${grantee}]: ${owner}`)
```

Comment:
Maybe `resolved grantee for managed grant [${grantee}]`?

Managed grant's grantee is the owner, "owner of a grantee" is confusing and that's not how we express relationships between actors in contracts and docs. Although, yes. From the perspective of this script's function, this is correct.